proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,692 @@
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+
+// New variant intended for clients sending the version key.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" value:"v1"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+```
+
+This will work fine for older clients that are not yet sending the
+`version` key, because their dynamic parameters will not match the new
+variant's constraints.  However, newer clients that are sending dynamic
+parameters `{env=prod, version=v1}` will run into ambiguity: those
+parameters can match either of the above variants of the resource.
+
+This situation will be avoided by requiring that **all variants of a
+given resource must specify constraints for the same set of keys**.
+
+However, in order to make this work for the case where the server starts
+sending the constraint on the new key before all clients are sending it,
+we provide the `exists` matcher, which will allow the server to specify
+a default explicitly for clients that are not yet sending a new key.
+In this example, the server would actually have the following two
+variants:
+
+```textproto
+// Existing variant for older clients that are not yet sending the
+// version key.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" exists:{} invert:true}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+
+// New variant for clients sending the version key.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" value:"v1"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+```
+
+This allows maintaining the requirement that all variants of a given
+resource have constraints on the same set of keys, while also allowing
+the server to explicitly provide a result for older clients that do not
+yet send the new key.
+
+##### Variants With Overlapping Constraint Values
+
+There is also a possible ambiguity that can occur if a server provides
+multiple variants of a resource whose constraints for a given key
+overlap in terms of the values they can match.  For example, let's say
+that a server has the following two variants of a resource:
+
+```textproto
+// Matches {env=prod} or {env=test}.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"env" value:"test"}
+    ]
+    match_type: MATCH_TYPE_OR
+  }
+]}
+
+// Matches {env=qa} or {env=test}.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"qa"},
+      {key:"env" value:"test"}
+    ]
+    match_type: MATCH_TYPE_OR
+  }
+]}
+```
+
+Now consider what happens if a client subscribes with dynamic parameters
+`{env=test}`.  Those dynamic parameters can match either of the above
+variants of the resource.
+
+This situation will be avoided by requiring that **all variants of a given
+resource must specify non-overlapping constraints for the same set of keys**.
+Control planes must not accept a set of resources that violates this
```

Comment:
This seems suspiciously similar to the 3-SAT NP-complete problem. A control plane could side-step that issue by giving a reduced configuration format and having limits, but is that what we had in mind? If so, we should provide some warning here "beware the computation complexity".

I honestly don't see the point of this requirement. The client already is supposed to handle ambiguity by selecting a resource arbitrarily. I could understand this requirement more if the client could assume there will never be ambiguity. Or if it is a "best practice" instead of "requirement."